sys/net/application_layer/sock_dodtls/sock_dodtls.c

```diff
@@ -0,0 +1,285 @@
+        send_duration = _now_ms() - start;
+        if (send_duration > CONFIG_SOCK_DODTLS_TIMEOUT_MS) {
+            return -ETIMEDOUT;
+        }
+        timeout -= send_duration;
+        if (res <= 0) {
+            _sleep_ms(timeout);
+            continue;
+        }
+        res = sock_dtls_recv(&_dtls_sock, &_server_session,
+                             _dns_buf, sizeof(_dns_buf), timeout);
+        if (res > 0) {
+            if (res > (int)SOCK_DODTLS_MIN_REPLY_LEN) {
+                uint32_t ttl = 0;
+
+                if ((res = dns_msg_parse_reply(_dns_buf, res, family,
+                                               addr_out, &ttl)) > 0) {
+                    dns_cache_add(domain_name, addr_out, res, ttl);
+                    goto out;
+                }
+            }
+            else {
+                res = -EBADMSG;
+            }
+        }
+    }
+
+out:
+    memset(_dns_buf, 0, sizeof(_dns_buf));  /* flush-out unencrypted data */
+    mutex_unlock(&_server_mutex);
+    return res;
+}
+
+int sock_dodtls_get_server(sock_udp_ep_t *server)
+{
+    int res = -ENOTCONN;
+
+    assert(server != NULL);
+    mutex_lock(&_server_mutex);
+    if (_server_set()) {
+        sock_udp_get_remote(&_udp_sock, server);
+        res = 0;
+    }
+    mutex_unlock(&_server_mutex);
+    return res;
+}
+
+sock_dtls_t *sock_dodtls_get_dtls_sock(void)
+{
+    return &_dtls_sock;
+}
+
+sock_dtls_session_t *sock_dodtls_get_server_session(void)
+{
+    return &_server_session;
+}
+
+int sock_dodtls_set_server(const sock_udp_ep_t *server,
+                           const credman_credential_t *creds)
+{
+    return (server == NULL)
+         ? _disconnect_server()
+         : _connect_server(server, creds);
+}
+
+static inline bool _server_set(void)
+{
+    return _cred_type != CREDMAN_TYPE_EMPTY;
+}
+
+static void _close_session(credman_tag_t creds_tag, credman_type_t creds_type)
+{
+    sock_dtls_session_destroy(&_dtls_sock, &_server_session);
+    sock_dtls_close(&_dtls_sock);
+    credman_delete(creds_tag, creds_type);
+    sock_udp_close(&_udp_sock);
+}
+
+static int _connect_server(const sock_udp_ep_t *server,
+                           const credman_credential_t *creds)
+{
+    int res = -EADDRINUSE;
+    uint32_t start, try_start, timeout = SOCK_DODTLS_SESSION_RECV_TIMEOUT_MS;
+    sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+
+    /* server != NULL is checked in sock_dodtls_set_server() */
+    assert(creds != NULL);
+    mutex_lock(&_server_mutex);
+    while (res == -EADDRINUSE) {
+        /* choose random ephemeral port, since DTLS requires a local port */
+        local.port = IANA_DYNAMIC_PORTRANGE_MIN +
+            (random_uint32() % (IANA_SYSTEM_PORTRANGE_MAX - IANA_DYNAMIC_PORTRANGE_MIN));
+        if ((res = sock_udp_create(&_udp_sock, &local, server, 0)) < 0) {
+            if (res != -EADDRINUSE) {
+                DEBUG("Unable to create UDP sock\n");
+                goto exit;
+            }
+        }
+    }
```

Comment:
Urgh - isn't this something `sock_udp_create()` should take care of when being called with `.port` set to 0?

Reply:
It does, however only on send. The issue here isn't `sock_udp`, but TinyDTLS's insistence to have a port assigned to the client socket (see first comment in the while loop).